contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
     return (countOfRemovalsAllocated, ids, amounts);
   }
 
+  /**
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplySpecificVintages(
+    uint256 amount,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    uint256 remainingAmountToFill = amount;
+    uint256 countOfListedRemovals = _removal.numberOfTokensOwnedByAddress({
+      account: address(this)
+    });
+    ids = new uint256[](countOfListedRemovals);
+    amounts = new uint256[](countOfListedRemovals);
+    suppliers = new address[](countOfListedRemovals);
+    countOfRemovalsAllocated = 0;
+    for (uint256 i = 0; i < countOfListedRemovals; ++i) {
+      uint256 removalId = _listedSupply[_currentSupplierAddress]
+        .getNextRemovalForSaleFromVintages({vintages: vintages});
+      // if removalId is 0 here, this supplier doesn't have any removals from the vintages requested
+      // at that point we should simply move on to the next supplier
+      if (removalId == 0) {
+        _incrementCurrentSupplierAddress();
+        continue;
+      }
+      uint256 removalAmount = _removal.balanceOf({
+        account: address(this),
+        id: removalId
+      });
+      if (remainingAmountToFill < removalAmount) {
+        /**
+         * The order is complete, not fully using up this removal, don't check about removing active supplier.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = remainingAmountToFill;
+        suppliers[countOfRemovalsAllocated] = _currentSupplierAddress;
+        remainingAmountToFill = 0;
```

Comment:
These if/else blocks seem like the could share more code, no? Especially if you can just do something like removalAmount - min(remainingAmountToFill, removalAmount)

Reply:
I was able to get better code reuse in all 4 allocation algorithms with this.  ran into some stack too deep errors and refactored how we manage `ids`, `amounts`, `suppliers`, and `countOfRemovalsAllocated` by creating the struct that contains them when they are initialized and referencing the struct throughout the algorithm.  This saves us a few lines of code, got rid of the stack too deep error, and I did a snapshot --diff to make sure the change didn't affect gas and there were only a few tests affected with changes less than 0.05% in all cases, with equal numbers of tests improving in gas as worsening.